File: beecrowd/Estruturas e bibliotecas/3163_ControladorDeVôo.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Celula{
    char valor[6];
    struct Celula* prox;
}Celula;

Celula* criaCelula(char s[]){
    Celula* c=malloc(sizeof(Celula));
    strcpy(c->valor,s);
    c->prox=NULL;
    return c;
}

typedef struct Fila{
    Celula* primeiro;
    Celula* ultimo;
}Fila;

Fila* criaFila(){
    Fila* f=malloc(sizeof(Fila));
    f->primeiro=NULL;
    f->ultimo=NULL;
    return f;
}

void inserir(Fila* f,char s[]){
    Celula* nova=criaCelula(s);
    if(f->primeiro==NULL){
        f->primeiro=nova;
        f->ultimo=nova;
        return;
    }
    f->ultimo->prox=nova;
    f->ultimo=nova;
}

void remover(Fila* f){
        if(f->primeiro==NULL)return;
        Celula* tmp=f->primeiro;
        f->primeiro=f->primeiro->prox;
        if(f->primeiro==NULL){
            f->ultimo=NULL;
        }
        free(tmp);
}

int isVazia(Fila* f){
    if(f->primeiro==NULL)return 1;
    
    return 0;
}

void mostrar(Fila* f){
    if(f->primeiro==NULL)return;
    Celula* aux=f->primeiro;
    while(aux!=NULL){
        if(aux->prox==NULL){
            printf("%s",aux->valor);
        }else{
            printf("%s ",aux->valor); 
        }
        aux=aux->prox;
    }
    printf("\n");
}

int main(){
    char p[6];

    Fila* norte=criaFila();
    Fila* sul=criaFila();
    Fila* oeste=criaFila();
    Fila* leste=criaFila();

    Fila* final=criaFila(); //fila fila na ordem certa

    scanf("%s",p);

    while(p[0]!='0'){
        char valor[6];
        if(p[0]=='-' && p[1]=='4'){ //leste
            scanf("%s",valor);
            while(valor[0]!='-' && valor[0]!='0'){
                inserir(leste,valor);
                scanf("%s",valor);
            }
            strcpy(p,valor);
            
        }else  if(p[0]=='-' && p[1]=='3'){ //norte
            scanf("%s",valor);
            while(valor[0]!='-' && valor[0]!='0' ){
                inserir(norte,valor);
                scanf("%s",valor);
            }
            strcpy(p,valor);
        }else  if(p[0]=='-' && p[1]=='2'){ //sul
            scanf("%s",valor);
            while(valor[0]!='-' && valor[0]!='0' ){
                inserir(sul,valor);
                scanf("%s",valor);
            }
            strcpy(p,valor);
        }else  if(p[0]=='-' && p[1]=='1'){ //oeste
            scanf("%s",valor);
            while(valor[0]!='-' && valor[0]!='0'){
                inserir(oeste,valor);
                scanf("%s",valor);
            }
            strcpy(p,valor);
        }
    }


    while(1){
        if(!isVazia(oeste)){
           inserir(final,oeste->primeiro->valor);
           remover(oeste);
        }
         if(!isVazia(norte)){
           inserir(final,norte->primeiro->valor);
           remover(norte);
        }
        if(!isVazia(sul)){
           inserir(final,sul->primeiro->valor);
           remover(sul);
        }
         if(!isVazia(leste)){
           inserir(final,leste->primeiro->valor);
           remover(leste);
        }
        if(isVazia(norte) && isVazia(sul) && isVazia(leste) && isVazia(oeste))break; //todos os listas vazias
    }

    mostrar(final);
    
    return 0;
}

File: beecrowd/Estruturas e bibliotecas/ 3163_ControladorDeVôo.java
import java.util.*;

class Aviao{
    String valor;
    Aviao prox;

    Aviao(String s){
        this.valor=s;
        this.prox=null;
    }
}

class Fila{
    Aviao primeiro,ultimo;

    Fila(){
        this.primeiro=null;
        this.ultimo=null;
    }

    void inserir(String s){
        Aviao novo=new Aviao(s);
        if(primeiro==null){

            primeiro=novo;
            ultimo=novo;
            return;
    }
    ultimo.prox=novo;
    ultimo=ultimo.prox;
}

void remover(){
    if(primeiro==null)return;
    Aviao tmp=primeiro;
    primeiro=primeiro.prox;
    tmp=null;
}

void mostrar(){
    Aviao a=primeiro;
    while(a!=null){

        if(a.prox==null){
        System.out.printf("%s\n",a.valor);
        }else{
        System.out.printf("%s ",a.valor);
        }
        a=a.prox;
    }

}

boolean isVazia(){
    return primeiro==null;
}

}

public class Main {
public static void main(String[] args){
        Scanner sc=new Scanner(System.in);
        String p=sc.next();
        Fila norte=new Fila();
        Fila sul=new Fila();
        Fila leste=new Fila();
        Fila oeste=new Fila();

        Fila fim=new Fila();

        while(!p.equals("0")){
            String valor;
            if(p.equals("-4")){
                valor=sc.next();
                while(valor.charAt(0)!='-' && valor.charAt(0)!='0'){
                    leste.inserir(valor);
                    valor=sc.next();
                }
                p=valor;
            }else if(p.equals("-3")){
                valor=sc.next();
                while(valor.charAt(0)!='-' && valor.charAt(0)!='0'){
                    norte.inserir(valor);
                    valor=sc.next();
                }
                p=valor;
            }else if(p.equals("-2")){
                valor=sc.next();
                while(valor.charAt(0)!='-' && valor.charAt(0)!='0'){
                    sul.inserir(valor);
                     valor=sc.next();
                }
                p=valor;
            }else if(p.equals("-1")){
                valor=sc.next();
                while(valor.charAt(0)!='-' && valor.charAt(0)!='0'){
                    oeste.inserir(valor);
                    valor=sc.next();
                }
                p=valor;
            }
        }
        
        while(!(norte.isVazia() && leste.isVazia() && oeste.isVazia() && sul.isVazia())){
                if(!oeste.isVazia()){
                    fim.inserir(oeste.primeiro.valor);
                    oeste.remover();
                }
                if(!norte.isVazia()){
                    fim.inserir(norte.primeiro.valor);
                    norte.remover();
                }
                if(!sul.isVazia()){
                    fim.inserir(sul.primeiro.valor);
                    sul.remover();
                }
                if(!leste.isVazia()){
                    fim.inserir(leste.primeiro.valor);
                    leste.remover();
                }
        } 
        fim.mostrar();

        sc.close();
    }
}